### DISCUSSION TITLE
Как сделать отдельное правило для локального порта?

### DISCUSSION THREAD

**Discussion by ivdeveloper86** [bol-van/zapret#293] [Q&A]
Всем привет!

У меня роутер ASUS, настроить получилось только через tpws.
Но вот телек LG никак не хочет работать, методом многочисkенных проб и ошибок подобрал параметр --mss=88 который дробит TCP пакеты и только так телек работает с Ютубом.
Естественно из-за этого все грузится крайне медленно на остальных устройствах сети, хочу разделить правила на основное и ТВ.
Файл dpi по дефолту
```
#!/bin/sh

SCRIPT=/tmp/mnt/DPI/zapret-master/tpws/tpws
PIDFILE=/var/run/tpws.pid
HOSTLISTFILE=/tmp/mnt/DPI/zapret-master/hostlist.txt
ARGS="--daemon --bind-addr 192.168.1.1 --port 998 --disorder --tlsrec=sni --split-pos=2 --pidfile $PIDFILE --hostlist $HOSTLISTFILE"

start() {
  if [ -f $PIDFILE ] && kill -0 $(cat $PIDFILE); then
    echo 'Service TPWS is already running' >&2
    return 1
  fi
  $SCRIPT $ARGS
  iptables -t nat -I PREROUTING -i br0 -p tcp --match multiport --dport 80,443 -j REDIRECT --to-port 998
  echo 'Started TPWS service'
}

stop() {
  if [ ! -f "$PIDFILE" ] || ! kill -0 $(cat "$PIDFILE"); then
    echo 'Service TPWS is not running' >&2
    return 1
  fi
  echo 'Stopping TPWS service...'
  kill -15 $(cat "$PIDFILE") && rm -f "$PIDFILE"
  iptables -t nat -D PREROUTING -i br0 -p tcp --match multiport --dport 80,443 -j REDIRECT --to-port 998
}

status() {
  if [ -f $PIDFILE ] && kill -0 $(cat $PIDFILE); then
    echo 'Service TPWS is running'
  else
    echo 'Service TPWS is stopped'
  fi
}

case "$1" in
  start)
    start
    ;;
  stop)
    stop
    ;;
  status)
    status
    ;;
  restart)
    stop
    start
    ;;
  *)
    echo "Usage: $0 {start|stop|restart|status}"
esac
```

Файл dpiTV для TV который подключен проводом в eth3
```
#!/bin/sh

SCRIPT=/tmp/mnt/DPI/zapret-master/tpws/tpws
PIDFILE=/var/run/tpws.pid
HOSTLISTFILE=/tmp/mnt/DPI/zapret-master/hostlist.txt
ARGS="--daemon --bind-addr 192.168.1.1 --port 999 --disorder --tlsrec=sni --split-pos=2 --pidfile $PIDFILE --mss=88 --hostlist $HOSTLISTFILE"

start() {
  if [ -f $PIDFILE ] && kill -0 $(cat $PIDFILE); then
    echo 'Service TPWS is already running' >&2
    return 1
  fi
  $SCRIPT $ARGS
  iptables -t nat -I PREROUTING -i eth3 -p tcp --match multiport --dport 80,443 -j REDIRECT --to-port 999
  iptables -I FORWARD -i eth3 -p udp --match multiport --dport 80,443 -j REJECT
  echo 'Started TPWS service'
}

stop() {
  if [ ! -f "$PIDFILE" ] || ! kill -0 $(cat "$PIDFILE"); then
    echo 'Service TPWS is not running' >&2
    return 1
  fi
  echo 'Stopping TPWS service...'
  kill -15 $(cat "$PIDFILE") && rm -f "$PIDFILE"
  iptables -t nat -D PREROUTING -i eth3 -p tcp --match multiport --dport 80,443 -j REDIRECT --to-port 999
  iptables -D FORWARD -i eth3 -p udp --match multiport --dport 80,443 -j REJECT
}

status() {
  if [ -f $PIDFILE ] && kill -0 $(cat $PIDFILE); then
    echo 'Service TPWS is running'
  else
    echo 'Service TPWS is stopped'
  fi
}

case "$1" in
  start)
    start
    ;;
  stop)
    stop
    ;;
  status)
    status
    ;;
  restart)
    stop
    start
    ;;
  *)
    echo "Usage: $0 {start|stop|restart|status}"
esac
```
Все это успешно стартует после ребута роутера, но похоже Телек все-равно работает по первому дефолтному правилу.
Подскажите что я делаю не так?

**Comment by kost2000** [comment]
Возможно лучше работать не с интерфейсом, а с ip адресом тв. Настроить телеку выдачу демоном DHCP статичного адреса и маршрутизировать этот адрес на вторую копию twps.

И ещё, должны быть разные разные имена фалов PIDFILE в скриптах.

**Comment by bol-van** [comment]
Хвататься на порт моста не получится
Пакеты будут приходить в br0, а не с eth3
Там надо фиксировать ip через dhcp или через статику и добавлять в правило -s 192.168.0.x

**Reply by ivdeveloper86** [reply]
Телеку назначен статический ИП адрес, можете написать полностью код?

**Reply by bol-van** [reply]
иптаблес такие
iptables -t nat -I PREROUTING -i br0 -p tcp -s 192.168.0.телик -match multiport --dport 80,443 -j REDIRECT --to-port 999
соответственно и -D тоже

НО

нельзя просто взять и запустить несколько приведенных выше скриптов с поправленым iptables
дело в очередности занесения правил. -I заносит в начало, опережая предыдущие
неизвестно в каком порядке они будут вызваны
так же нельзя использовать конфликтующие параметры вроде одного и того же pidfile или номера порта

редирект с -s должен в правилах стоять выше общего редиректа
значит если есть 2 команды iptables -I, то специфический редирект должен быть ниже по тексту

**Reply by ivdeveloper86** [reply]
То есть дефолтный файл iptables -t nat -A
Файл ТВ iptables -t nat -I
Так нужно сделать?

**Reply by bol-van** [reply]
это все надо засовывать в один скрипт, немного его переписав
иначе будут гонки

**Reply by ivdeveloper86** [reply]
так весь смысл, чтобы --mss не было в основном скрипте или это можно как-то разделить в одном скрипте?

**Reply by bol-van** [reply]
скрипт это просто батник. запуск чего сделаете то и будет.
инстансы тпвс разделяются по портам. их может быть много
хотя если совсем не соображаете в программировании может быть непросто
но копипаста и дублирование скрипта тоже не поможет
поймите как работает скрипт и вопросы отпадут

**Reply by ivdeveloper86** [reply]
пока оставлю вариант с 2 скриптами по вашим подсказкам, вроде все работает
